micro-rdk/src/common/data_collector.rs

```diff
@@ -0,0 +1,352 @@
+pub enum CollectionMethod {
+    Readings,
+    // MovementSensor methods
+    AngularVelocity,
+    LinearAcceleration,
+    LinearVelocity,
+    // TODO: PowerSensor methods
+    // Voltage,
+    // Current,
+    // TODO: Board
+    // Analogs(String)
+}
+
+impl CollectionMethod {
+    fn method_str(&self) -> String {
+        match self {
+            Self::Readings => "readings".to_string(),
+            Self::AngularVelocity => "angularvelocity".to_string(),
+            Self::LinearAcceleration => "linearacceleration".to_string(),
+            Self::LinearVelocity => "linearvelocity".to_string(),
+            // TODO: Power Sensor and Board collectors
+            // Self::Voltage => "voltage".to_string(),
+            // Self::Current => "current".to_string(),
+            // Self::Analogs(_) => "analogs".to_string()
+        }
+    }
+}
+
+impl Display for CollectionMethod {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(&self.method_str(), f)
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum DataCollectionError {
+    #[error("method {0} unsupported for {1}")]
+    UnsupportedMethod(CollectionMethod, String),
+    #[error("no collection methods supported for component")]
+    NoSupportedMethods,
+    #[error(transparent)]
+    SensorCollectionError(#[from] SensorError),
+    // TODO: remove when error enums for all components are created (RSDK-6909)
+    #[error(transparent)]
+    MiscCollectionError(#[from] anyhow::Error),
+}
+
+/// A DataCollector represents an association between a data collection method and
+/// a ResourceType (i.e. SensorType & Readings, BoardType & Analogs) and the frequency at
+/// which the results of the method should be stored. It is capable
+pub struct DataCollector {
+    name: String,
+    component_type: String,
+    resource: ResourceType,
+    method: CollectionMethod,
+    time_interval_ms: u64,
+}
+
+fn resource_method_pair_is_valid(resource: &ResourceType, method: &CollectionMethod) -> bool {
+    match resource {
+        ResourceType::Sensor(_) => matches!(method, CollectionMethod::Readings),
+        ResourceType::MovementSensor(_) => matches!(
+            method,
+            CollectionMethod::Readings
+                | CollectionMethod::AngularVelocity
+                | CollectionMethod::LinearAcceleration
+                | CollectionMethod::LinearVelocity
+        ),
+        // ResourceType::PowerSensor(_) => matches!(method, CollectionMethod::Readings | CollectionMethod::Voltage | CollectionMethod::Current),
+        // ResourceType::Board(_) => matches!(method, CollectionMethod::Analogs(_)),
+        _ => false,
+    }
+}
+
+impl DataCollector {
+    pub fn new(
+        name: String,
+        resource: ResourceType,
+        method: CollectionMethod,
+        capture_frequency_hz: f32,
+    ) -> Result<Self, DataCollectionError> {
+        let time_interval_ms = ((1.0 / capture_frequency_hz) * 1000.0) as u64;
```

Comment:
Or just `1000.0 / capture_frequence_hz as u64`

Reply:
yeah my bad, wrote the way I think and then forgot to clean up